File: backend/csharp-runner/src/execute-tests.ts
import { promisify } from 'util';
import { exec } from 'child_process';
import { readFile,mkdtemp} from 'fs/promises';
import * as fs from 'fs';
const ncp = require('ncp').ncp;

export async function runCSharp(exerciseName: string, templateFilePath:string,code:string, fileName: string): Promise<any> {
  const solutionDir = await createTempDirAndCopyTemplate(exerciseName, templateFilePath);
  await replaceCode(code, solutionDir,fileName);

  await runCommands(`/usr/src/app/${solutionDir}`, `dotnet restore`);
  await runCommands(`/usr/src/app/${solutionDir}`, `dotnet test -l:trx;LogFileName=TestOutput.xml`);

  const result = await readFile(`/usr/src/app/${solutionDir}/results/testresults.json`, 'utf-8');
  return JSON.parse(result);
}

async function replaceCode(code: string, filePath: string, fileName:string): Promise<void> {
  const templateFilePath = `/usr/src/app/${filePath}/src/${fileName}`;

  return new Promise<void>((resolve, reject) => {
    fs.writeFile(templateFilePath, code, (err: any) => {
      if (err) {
        console.error(err);
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

async function runCommands(path: string, command: string): Promise<void> {
  try {
    const { stdout, stderr } = await promisify(exec)(command, { cwd: path });
  } catch (error: any) {
    console.error(error.message);
  }
}


async function createTempDirAndCopyTemplate(exerciseName: string, templateFilePath: string): Promise<string> {
  let solutionDir = await mkdtemp(exerciseName);

  await new Promise<void>((resolve, reject) => {
    ncp(templateFilePath, solutionDir, { clobber: false }, function (err: any) {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });

  return solutionDir;
}

File: backend/csharp-runner/csharp-runner/executeTests.cs
using System.Diagnostics;

namespace csharp_runner
{
    public class executeTests
    {
        public static async Task<string> runCSharp(string exerciseName, string templateFilePath, string code, string fileName)
        {
            string solutionDir = createTempDirAndCopyTemplate(templateFilePath, exerciseName);

            replaceCode(solutionDir, code, fileName);

            runCommands(solutionDir, "restore");
            runCommands(solutionDir, "test -l:trx;LogFileName=TestOutput.xml");

            string testOutput = await File.ReadAllTextAsync(Path.Combine(solutionDir, "TestOutput.xml"));
            Console.WriteLine(testOutput);
            return testOutput;
        }

        private static void runCommands(string solutionDir, string command)
        {
            ProcessStartInfo psi = new ProcessStartInfo
            {
                FileName = "dotnet",
                Arguments = $"{command}",
                WorkingDirectory = solutionDir,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using (Process process = new Process { StartInfo = psi })
            {
                process.Start();
                string output = process.StandardOutput.ReadToEnd();
                string error = process.StandardError.ReadToEnd();
                process.WaitForExit();
            }
        }

        private static async void replaceCode(string solutionDir, string code, string fileName)
        {
            await File.WriteAllTextAsync(Path.Combine(solutionDir, fileName), code);
        }

        private static string createTempDirAndCopyTemplate(string templateFilePath, string exerciseName)
        {
            string solutionDir = Directory.CreateTempSubdirectory().Name;
            File.Copy(templateFilePath, solutionDir);
            return solutionDir;
        }
    }
}